heron/api/src/scala/Streamlet.scala

```diff
@@ -0,0 +1,231 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// Copyright 2016 Twitter. All rights reserved.
+package com.twitter.heron.streamlet.scala
+import java.util.List
+import com.twitter.heron.classification.InterfaceStability
+
+
+/**
+  * A Streamlet is a (potentially unbounded) ordered collection of tuples.
+  * Streamlets originate from pub/sub systems(such Pulsar/Kafka), or from
+  * static data(such as csv files, HDFS files), or for that matter any other
+  * source. They are also created by transforming existing Streamlets using
+  * operations such as map/flatMap, etc.
+  * Besides the tuples, a Streamlet has the following properties associated with it
+  * a) name. User assigned or system generated name to refer the streamlet
+  * b) nPartitions. Number of partitions that the streamlet is composed of. Thus the
+  *    ordering of the tuples in a Streamlet is wrt the tuples within a partition.
+  *    This allows the system to distribute  each partition to different nodes across the cluster.
+  * A bunch of transformations can be done on Streamlets(like map/flatMap, etc.). Each
+  * of these transformations operate on every tuple of the Streamlet and produce a new
+  * Streamlet. One can think of a transformation attaching itself to the stream and processing
+  * each tuple as they go by. Thus the parallelism of any operator is implicitly determined
+  * by the number of partitions of the stream that it is operating on. If a particular
+  * transformation wants to operate at a different parallelism, one can repartition the
+  * Streamlet before doing the transformation.
+  */
+@InterfaceStability.Evolving
+trait Streamlet[R] {
+
+  /**
+    * Sets the name of the BaseStreamlet.
+    * @param sName The name given by the user for this BaseStreamlet
+    * @return Returns back the Streamlet with changed name
+    */
+  def setName(sName: String): Streamlet[R]
+
+  /**
+    * Gets the name of the Streamlet.
+    * @return Returns the name of the Streamlet
+    */
+  def getName(): String
+
+  /**
+    * Sets the number of partitions of the streamlet
+    * @param numPartitions The user assigned number of partitions
+    * @return Returns back the Streamlet with changed number of partitions
+    */
+  def setNumPartitions(numPartitions: Int): Streamlet[R]
+
+  /**
+    * Gets the number of partitions of this Streamlet.
+    * @return the number of partitions of this Streamlet
+    */
+  def getNumPartitions(): Int
+
+  /**
+    * Return a new Streamlet by applying mapFn to each element of this Streamlet
+    * @param mapFn The Map Function that should be applied to each element
+    */
+  def map[T](mapFn: SerializableFunction[R, _ <: T]): Streamlet[T]
```

Comment:
In the light of my previous comment, as a sample for `map` function, i think this should be as follows.
`def map[T](mapFn: R => _ <: T): Streamlet[T]`
instead of 
`def map[T](mapFn: SerializableFunction[R, _ <: T]): Streamlet[T]`

Same comment for the following functions.